### DISCUSSION TITLE
Translating to Python

### DISCUSSION THREAD

**Discussion by christj99** [Alkaar/resy-booking-bot#113] [Q&A]
Going out on a limb and hoping someone who knows both python and Scala well enough can help. I was running python code that basically mirrored this Scala code as I'm not very familiar with Scala. Both ran well until the issue of the missing origin header in the book POST request came up #104 . Adding the origin header seemed to fix it here but not in the python code for some reason. To simplify things, I've provided just the code that would go into the post request: 

```
import requests

headers = {'Authorization': 'ResyAPI api_key="<my key>"', 
           'x-resy-auth-token': '<my token>', 
           'origin': 'https://widgets.resy.com/', #This was added after the fix was made in the scala code
           'Content-Type': 'application/x-www-form-urlencoded'
           }

params = {'book_token': 'Q_vA|3Dhi4E_pudZ|IfGlTszU3wk7Ni17N8I09JFosu5FAINi51OG2vSU_P8EIFiWHAoGLT0Tr5yAhy7TBX_qnGz4CTTfZ7NdV4D7nVK_JukUtcJ90cgGsO1ZXvL8MdzKmwDXoSjO0eBejMtO1bTZavIf6ma7c75owHBKirWjJwj4scmBis9zIYiaS_AJo_QTBUr9wYsYgeo2R|MMgqMo_|NfLYU6NAT_31Cc4Qtofvq7kWbl24A932lZr9OYxAACrWNBBOl4NiEmAxWCgiCwOx2SB4ez1AvdFPfOXv93V7kBdMZVWNXbtMIvpK2g|A5LX7YNcrwk6TrqSurOnYf|GyNUVSiiD5FbJr4IJEuyxA5Qy1NgqfZ7vtyqJbKnxgEBspgg4gLohskOhq2_BS0sl_HVPmZqGSIqmCHjJmVjeE4qHeYo0g9CrmPYg3IxDZRiTUAQ1kkVSIdSJ6tUUCrSlLtlk8x1R1YmrdLKOx5Av3Ot4I5QVN4yUsOdQQCUsmgUe_tD0k1tw0i|85IRVcxaCxBpNpGM|P6jY7JJLFYM5_QPkglJceMUFimT25_|8DYDps6v2rjIVtyKyFEyzS1Osp2U0bDYAJVCX19a_58YstiAEf5xHqBf1FA1XIwRKgXm0nNrkkdtQBWMgS6ko|aPQIbHrBneG2R4vy|_|tgqkUDMaduF|OHFTX29yFpSzRu-7e418a144e9ecd73a6d20715b16976548ec8e4462e47d56a7549e638', 
          'struct_payment_method': "{'id': <my id>}"}

response = requests.post('https://api.resy.com/3/book', headers=headers, params=params)
```
This worked up until a couple of weeks ago, and then broke like everyone else's code when #102 happened. However, adding the origin header didn't seem to fix it like it did here, and I still get the 500 error code. 

Any idea's as to what I'm missing here? something getting lost in translation? I'm struggling given that mirroring the Scala code has worked up to this point however I don't know enough Scala to know what I'm missing this time.

**Comment by ahammer3** [comment]
I am attempting this myself, and am at the same point as you. Put together some skeleton code that worked up until a few weeks ago when the orgin header issue popped out, and I'm also getting a 500 error code now. Everything works up until the book API call. If I find a solution or make any progress I'll post an update here.

**Reply by christj99** [reply]
Thanks! I'll keep looking and do the same.

**Reply by ryantorsparks** [reply]
I'm also doing the same - I think if you add ```'pragma': 'no-cache'``` to the booking headers it succeeds for restaurants that don't require payment methods (no idea why that works). However, I can't get it to work for restaurants that require payment methods, for those I get
```
>>> print(response.text)
{"status": 400, "code": null, "message": "Invalid data received.", "data": {"struct_payment_method": "invalid"}}
```
My inputs are


```
    book_headers = {
    'origin': 'https://widgets.resy.com',
    'Content-Type': 'application/x-www-form-urlencoded',
    'pragma': 'no-cache',
    'Authorization':  resyApiKey,
    'x-resy-auth-token': userAuthToken
    }

bookURL='https://api.resy.com/3/book'

book_data= {
    'book_token' : bookToken,
    'struct_payment_method' : {'id' : <my payment id, int > }
    }

```

Not sure if anyone has any further ideas? The exact same request seems to work fine with scala

**Comment by Alkaar** [comment]
I just pushed fix https://github.com/Alkaar/resy-booking-bot/pull/122 where it seems like the Referer header is now required. Try adding that and see if that addresses your issue.

**Reply by christj99** [reply]
Unfortunately, it didn't work but thanks for the help! it's useful for narrowing down whatever the culprit could be. How do you go about knowing what headers are needed when the code begins to have trouble?

**Comment by clejcode** [comment]
Has anyone had success with a full Python implementation? Just starting to look into this now and am also way more comfortable with Python than Scala
